File: go-side/internal/layers/application/application.go
package application

import (
	"bufio"
	"fmt"
	"io"
	"strings"

	"cc3067/lab2/go-side/internal/layers/link"
)

type OutgoingRequest struct {
	Message         string
	Algorithm       string
	ProbabilityText string
}

func readLine(reader *bufio.Reader, writer io.Writer, prompt string) (string, error) {
	fmt.Fprint(writer, prompt)
	line, err := reader.ReadString('\n')
	if err != nil && len(line) == 0 {
		return "", err
	}
	return strings.TrimRight(line, "\r\n"), nil
}

func SolicitarMensaje(reader *bufio.Reader, writer io.Writer) (OutgoingRequest, error) {
	message, err := readLine(reader, writer, "Mensaje ASCII (o /salir): ")
	if err != nil {
		return OutgoingRequest{}, err
	}
	if message == "/salir" {
		return OutgoingRequest{}, io.EOF
	}
	algorithmText, err := readLine(reader, writer, "Algoritmo [1=hamming, 2=crc32]: ")
	if err != nil {
		return OutgoingRequest{}, err
	}
	algorithm, err := link.NormalizeAlgorithm(algorithmText)
	if err != nil {
		return OutgoingRequest{}, err
	}
	probability, err := readLine(reader, writer, "Probabilidad de error [decimal o fraccion, ej. 1/100]: ")
	if err != nil {
		return OutgoingRequest{}, err
	}
	return OutgoingRequest{message, algorithm, probability}, nil
}

func MostrarMensaje(message, errorMessage string, corrected bool) {
	if errorMessage != "" {
		fmt.Printf("\n[APLICACION] ERROR: %s\n", errorMessage)
		return
	}
	suffix := ""
	if corrected {
		suffix = " (Hamming corrigio un bit)"
	}
	fmt.Printf("\n[APLICACION] Mensaje recibido%s: %s\n", suffix, message)
}

File: proyecto/python-side/atm/layers/application.py
"""Servicios de la capa de Aplicacion."""

from dataclasses import dataclass
from typing import Callable

from atm.layers.link import normalize_algorithm


@dataclass(frozen=True)
class OutgoingRequest:
    """Datos capturados por Aplicacion antes de atravesar las otras capas."""

    message: str
    algorithm: str
    probability_text: str


def solicitar_mensaje(input_fn: Callable[[str], str] = input) -> OutgoingRequest:
    """Solicita texto, algoritmo y ruido sin codificar ni transmitir datos."""
    message = input_fn("Mensaje ASCII (o /salir): ")
    if message == "/salir":
        raise EOFError
    algorithm = normalize_algorithm(input_fn("Algoritmo [1=hamming, 2=crc32]: "))
    probability = input_fn("Probabilidad de error [decimal o fraccion, ej. 1/100]: ")
    return OutgoingRequest(message, algorithm, probability)


def mostrar_mensaje(message: str | None = None, error: str | None = None, corrected: bool = False) -> None:
    """Presenta al usuario una entrega valida o el error reportado por las capas."""
    if error:
        print(f"\n[APLICACION] ERROR: {error}")
        return
    suffix = " (Hamming corrigio un bit)" if corrected else ""
    print(f"\n[APLICACION] Mensaje recibido{suffix}: {message}")
